### DISCUSSION TITLE
How to match chained expressions

### DISCUSSION THREAD

**Discussion by emosenkis** [ast-grep/ast-grep#801] [Q&A] [answered]
I'm trying to add type annotations to SQLAlchemy queries. In this simplified [example](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiQCRQQVJFTlQuZmllbGQoJEZJRUxEKVxuZGVmICRGVU5DKCQkJEFSR1MpIC0+ICRSRVRVUk46XG4gICQkJEJPRFkiLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoicnVsZTpcbiAga2luZDogYXNzaWdubWVudFxuICBwYXR0ZXJuOiAkVkFSID0gJFZBTFxuICBoYXM6XG4gICAga2luZDogY2FsbFxuICAgIHBhdHRlcm46IGRiLnNlc3Npb24ucXVlcnkobW9kZWxzLiRUWVBFKVxuICAgIHN0b3BBdDogZW5kXG5maXg6ICckVkFSOiBtb2RlbHMuJFRZUEUgPSAkVkFMJyIsInNvdXJjZSI6ImltcG9ydCBkYlxuaW1wb3J0IG1vZGVsc1xuXG5mb28gPSBkYi5zZXNzaW9uLnF1ZXJ5KG1vZGVscy5Gb28pLmZpbHRlcihUcnVlKS5hbGwoKSJ9), although in the tree-sitter output I can see a `call` node with the content `db.session.query(models.Foo)`, there is no match. I've also been completely unable to match directly on the complete query with patterns like `db.session.query($TYPE).$$$REST` (to be really useful, I would also need to be able to match on the suffix since the type is `Query[$TYPE]` in the simple case but simply `$TYPE` if it ends in `.one()`, `$TYPE | NONE` if it ends in `.first()`, or `list[$TYPE]` if it ends in `.all()`

**Comment by HerringtonDarkholme** [comment, marked as answer]
There are multiple ways to achieve it.


1. Using has operator and conditional stopBy 
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiQCRQQVJFTlQuZmllbGQoJEZJRUxEKVxuZGVmICRGVU5DKCQkJEFSR1MpIC0+ICRSRVRVUk46XG4gICQkJEJPRFkiLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoicnVsZTpcbiAga2luZDogY2FsbFxuICBwYXR0ZXJuOiAkQ0hBSU5cbiAgaGFzOlxuICAgIHBhdHRlcm46IGRiLnNlc3Npb24ucXVlcnkoJEIpXG4gICAgc3RvcEJ5OlxuICAgICAgbm90OlxuICAgICAgICBhbnk6XG4gICAgICAgICAgLSBraW5kOiBjYWxsIFxuICAgICAgICAgIC0ga2luZDogYXR0cmlidXRlIiwic291cmNlIjoiaW1wb3J0IGRiXG5pbXBvcnQgbW9kZWxzXG5cbmZvbyA9IGRiLnNlc3Npb24ucXVlcnkobW9kZWxzLkZvbykuZmlsdGVyKFRydWUpLmFsbCgpIn0=

2. use inside operator and not inside rule 
Note, you need to extract the full match from the Pattern `$B`
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJERCLmFsbCgpIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIHBhdHRlcm46IGRiLnNlc3Npb24ucXVlcnkoJEEpXG4gIGluc2lkZTogXG4gICAgYWxsOlxuICAgICAgLSBraW5kOiBjYWxsXG4gICAgICAtIG5vdDpcbiAgICAgICAgICBpbnNpZGU6IFxuICAgICAgICAgICAga2luZDogYXR0cmlidXRlXG4gICAgICAtIHBhdHRlcm46ICRCXG4gICAgc3RvcEJ5OiBlbmQiLCJzb3VyY2UiOiJpbXBvcnQgZGJcbmltcG9ydCBtb2RlbHNcblxuZm9vID0gZGIuc2Vzc2lvbi5xdWVyeShtb2RlbHMuRm9vKS5maWx0ZXIoVHJ1ZSkuYWxsKClcblxudGVzdChkYi5zZXNzaW9uLnF1ZXJ5KDEyMykudGVzdCgpKVxuXG5cbnRlc3QuY2FsbCgpLnNzcygpXG5cbnRlc3QuY2FsbChkYi5zZXNzaW9uLnF1ZXJ5KG1vZGVscy5CYXIpKSJ9

3. Use utility rule to recursively find all attributes
https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiJERCLmFsbCgpIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InV0aWxzOlxuICBEQl9DQUxMOlxuICAgIGFueTpcbiAgICAgIC0gcGF0dGVybjogJ2RiLnNlc3Npb24ucXVlcnkoJFRZUEUpJ1xuICAgICAgLSBraW5kOiBjYWxsXG4gICAgICAgIGhhczpcbiAgICAgICAgICBraW5kOiBhdHRyaWJ1dGVcbiAgICAgICAgICBoYXM6IHsgbWF0Y2hlczogREJfQ0FMTCB9XG5ydWxlOlxuICBtYXRjaGVzOiBEQl9DQUxMIiwic291cmNlIjoiaW1wb3J0IGRiXG5pbXBvcnQgbW9kZWxzXG5cbmZvbyA9IGRiLnNlc3Npb24ucXVlcnkobW9kZWxzLkZvbykuZmlsdGVyKFRydWUpLmFsbCgpXG5cbnRlc3QoZGIuc2Vzc2lvbi5xdWVyeSgxMjMpLnRlc3QoKSlcblxuXG50ZXN0LmNhbGwoKS5zc3MoKVxuXG50ZXN0LmNhbGwoZGIuc2Vzc2lvbi5xdWVyeShtb2RlbHMuQmFyKSkifQ==

**Reply by emosenkis** [reply]
Thanks! The first suggestion worked perfectly! I ended up doing this (with several variants for different terminal calls)

```
id: typed-query-all
language: py
rule:
  kind: assignment
  pattern: $VAR = $VAL
  regex: '\.all\(\)\s*\)*$'
  has:
    pattern: db.session.query($B)
    stopBy:
      not:
        any:
          - kind: parenthesized_expression
          - kind: call
          - kind: attribute

fix: '$VAR: list[$B] = $VAL'
```
